### ISSUE TITLE
Fail to build from local SHA-ID with DOCKER_BUILDKIT=1

### ISSUE THREAD

**Issue by virgile-hogman** [moby/moby#39769]
**Description**

Cannot build with `DOCKER_BUILDKIT=1` from a local image when using a local SHA-ID (`FROM sha256:xxxx`) instead of an image name. This works without `DOCKER_BUILDKIT=1`. This applies to a multi-stage setup or simply from a local image in a CI setup.

**Steps to reproduce the issue:**
1. Demo project available here: https://github.com/virgile-hogman/docker-buildkit
2. Run `./docker-build.sh` from bash
3. See logs, fails to resolve `FROM docker.io/library/sha256:xxxx`

**Describe the results you received:**

+ DOCKER_BUILDKIT=1
+ docker build -t bk2 --no-cache --build-arg BASE_IMAGE=sha256:61e3fb57fa55b8ebf1a944c55f237740f4941f675551de64ea4199efecf3ee8a src-custom
[+] Building 2.4s (4/5)                                                                                                
 => [internal] load build definition from Dockerfile                                                              0.1s
 => => transferring dockerfile: 37B                                                                               0.0s
 => [internal] load .dockerignore                                                                                 0.1s
 => => transferring context: 2B                                                                                   0.0s
 => ERROR [internal] load metadata for docker.io/library/sha256:61e3fb57fa55b8ebf1a944c55f237740f4941f675551de64  1.8s
 => ERROR [1/2] FROM docker.io/library/sha256:61e3fb57fa55b8ebf1a944c55f237740f4941f675551de64ea4199efecf3ee8a    0.5s
 => => resolve docker.io/library/sha256:61e3fb57fa55b8ebf1a944c55f237740f4941f675551de64ea4199efecf3ee8a          0.5s
------
 > [internal] load metadata for docker.io/library/sha256:61e3fb57fa55b8ebf1a944c55f237740f4941f675551de64ea4199efecf3ee8a:
------
------
 > [1/2] FROM docker.io/library/sha256:61e3fb57fa55b8ebf1a944c55f237740f4941f675551de64ea4199efecf3ee8a:
------
pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed

**Describe the results you expected:**

This should work as for `docker build` without `DOCKER_BUILDKIT=1`.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:           18.09.6
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        481bc77
 Built:             Sat May  4 02:35:57 2019
 OS/Arch:           linux/amd64
 Experimental:      false
```

**Output of `docker info`:**

```
Does not seem relevant for now, can provide if really needed.
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Native install with Ubuntu 18.04. Connected to a private repository on a different host but this should not have effect here since it works with standard `docker build`.

**Comment by thaJeztah** [comment]
ping @tonistiigi @AkihiroSuda ptal

**Comment by thaJeztah** [comment]
To prevent confusion; the `sha256` here is the image-ID, not the image's digest;

```bash
docker images --digests --no-trunc --filter reference=busybox
REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID                                                                  CREATED             SIZE
busybox             latest              sha256:9f1003c480699be56815db0f8146ad2e22efea85129b5b5983d0e0fb52d9ab70   sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb   4 weeks ago         1.22MB
```

Still reproducible on Docker 19.03.1;

### Buildkit disabled

```bash
DOCKER_BUILDKIT=0 docker build --no-cache -<<EOF
FROM sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb
RUN echo hello
EOF
```

```
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb
 ---> db8ee88ad75f
Step 2/2 : RUN echo hello
 ---> Running in b68b589b70f3
hello
Removing intermediate container b68b589b70f3
 ---> e7294dc9a687
Successfully built e7294dc9a687
```

```
DOCKER_BUILDKIT=1 docker build --no-cache -<<EOF
FROM sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb
RUN echo hello
EOF
```

### Buildkit enabled

```
[+] Building 1.5s (4/5)                                                                                                                                                                                                                                                      
 => [internal] load .dockerignore                                                                                                                                                                                                                                       0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                         0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                    0.0s
 => => transferring dockerfile: 134B                                                                                                                                                                                                                                    0.0s
 => ERROR [internal] load metadata for docker.io/library/sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb                                                                                                                                        1.1s
 => ERROR [1/2] FROM docker.io/library/sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb                                                                                                                                                          0.4s
 => => resolve docker.io/library/sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb                                                                                                                                                                0.4s
------
 > [internal] load metadata for docker.io/library/sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb:
------
------
 > [1/2] FROM docker.io/library/sha256:db8ee88ad75f6bdc74663f4992a185e2722fa29573abcc1a19186cc5ec09dceb:
------
pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

**Comment by virgile-hogman** [comment]
> To prevent confusion; the sha256 here is the IMAGE ID, not the image's digest;

Precisely! I didn't try with the digest though, but it's less relevant for me. My use case is that i generate a local image in a CI setup without any image name. The reason for this is that several jobs can be triggered at the same time by multiple processes so dealing with the SHA only solves any potential conflict.

It's not entirely clear is there is a precise specification of a repo but to me it looks like `registry/user/basename:tag@sha256:id`. The nice thing with `docker build` is its flexibility, leaving almost any field optional with logic heuristic rules. Are they specified somewhere? Not even sure about it.

For instance the image@sha could disambiguate though i don't even know in practice if the same image:tag can have multiple SHA at the same time. But without name the @ is not even necessary though the `sha256:` prefix certainly helps to disambiguate :smile:

**Comment by thaJeztah** [comment]
> I don't even know in practice if the same image:tag can have multiple SHA at the same time. But without name the @ is not even necessary though the sha256: prefix certainly helps to disambiguate 😄

The notation `image:tag@sha256:...` is a bit ambiguous; it's mainly to give users a visible "clue" what tag was used originally to resolve the digest.

A tag can only be associated with a single digest at any point, i.e. `busybox:latest@sha256:9f1003c480699be56815db0f8146ad2e22efea85129b5b5983d0e0fb52d9ab70` is true _today_ (with the output in my previous example), but tomorrow, `busybox:latest` may resolve to a different digest.

Docker actually ignores the `:tag` if a digest is provided, so `image:tag@<digest>` is equivalent to `image@digest` (the digest is leading, because it's immutable). The difficult bit (today) is that the digest is only calculated when pushing an image to a registry, which (in your use-case where you don't push the image) probably won't be helpful 😓

I _think_ it's possible to calculate the digest _without_ actually pushing (containerd may already do this, but @tonistiigi may be able to confirm that)

**Comment by mdeknowis** [comment]
Perhaps it helps to know: I was using `DOCKER_BUILDKIT=1` in a `sh` script. After migrating it into a `bash` script I experienced wired issues. So I've to invoke `sh docker build .`, then everything is fine again. Hence it looks like the  `DOCKER_BUILDKIT=1` does not work fine in `bash`. But very well in `sh`

**Comment by virgile-hogman** [comment]
Any news about this?

> Perhaps it helps to know: I was using `DOCKER_BUILDKIT=1` in a `sh` script. After migrating it into a `bash` script I experienced wired issues. So I've to use `sh`, then everything is fine again. Hence it looks like the `DOCKER_BUILDKIT=1` does not work fine in `bash`. But very well in `sh`

@mdeknowis I don't see how the shell could impact this issue. What do you mean with "wired issues"? Please be precise.

**Comment by orgads** [comment]
As a workaround, you can create a tag before building, and use it in the `FROM` statement.

**Comment by zapotocnylubos** [comment]
Build work with disabled buildkit
`FROM sha256:<ImageID>`
Why is this not supported in buldkit?
